Assets/BossRoom/Scripts/Shared/Net/NetworkObjectPool.cs

```diff
@@ -0,0 +1,207 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using MLAPI;
+using MLAPI.Spawning;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace BossRoom.Scripts.Shared.Net.NetworkObjectPool
+{
+    public class NetworkObjectPool : MonoBehaviour
+    {
+        private static NetworkObjectPool _instance;
+
+        public static NetworkObjectPool Singleton { get { return _instance; } }
+
+        [SerializeField]
+        NetworkManager m_NetworkManager;
+
+        [SerializeField]
+        List<PoolConfigObject> PooledPrefabsList;
+
+        HashSet<GameObject> prefabs = new HashSet<GameObject>();
+
+        Dictionary<GameObject, Queue<NetworkObject>> pooledObjects = new Dictionary<GameObject, Queue<NetworkObject>>();
+
+        private bool m_HasInitialized = false;
+
+        public void Awake()
+        {
+            if (_instance != null && _instance != this)
+            {
+                Destroy(this.gameObject);
+            } else {
+                _instance = this;
+                DontDestroyOnLoad(this);
```

Comment:
I wonder whether we would really want to keep this alive when we leave the game scene. I guess it's fine here but usually I'd destroy pools like that and clear resources.